server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -0,0 +1,240 @@
+
+  // It takes at least 5 seconds to see the newly submitted app.
+  private val APP_TAG_TO_ID_TIMEOUT = 30 seconds
+  private val KILL_TIMEOUT = APP_TAG_TO_ID_TIMEOUT
+  private val POLL_INTERVAL = 1 second
+
+  // YarnClient is thread safe. Create once, share it across threads.
+  private[this] lazy val yarnClient = YarnClient.createYarnClient()
+  private[this] var yarnClientCreated = false
+
+  def getYarnClient(): YarnClient = synchronized {
+    if (!yarnClientCreated) {
+      yarnClient.init(new YarnConfiguration())
+      yarnClient.start()
+      yarnClientCreated = true
+    }
+    yarnClient
+  }
+
+  def getAppIdFromTagAsync(appTag: String): Future[ApplicationId] =
+    Future { getAppIdFromTag(appTag).get }
+
+  /**
+   * Find the corresponding YARN application id from an application tag.
+   *
+   * @param appTag The application tag tagged on the target application.
+   *               If the tag is not unique, it returns the first application it found.
+   *               It will be converted to lower case to match YARN's behaviour.
+   * @return ApplicationId or the failure.
+   */
+  @tailrec
+  private def getAppIdFromTag(
+    appTag: String,
+    deadline: Deadline = APP_TAG_TO_ID_TIMEOUT.fromNow): Try[ApplicationId] = {
+    val appTagLowerCase = appTag.toLowerCase
+
+    // FIXME Should not loop thru all YARN applications but YarnClient doesn't offer an API.
+    // Consider calling rmClient in YarnClient directly.
+    getYarnClient().getApplications().asScala.find(_.getApplicationTags.contains(appTagLowerCase))
+      match {
+        case Some(app) => Success(app.getApplicationId)
+        case None =>
+          if (deadline.isOverdue) {
+            Failure(new Exception(s"No YARN application is tagged with $appTagLowerCase."))
+          } else {
+            blocking { Clock.sleep(POLL_INTERVAL.toMillis) }
+            getAppIdFromTag(appTagLowerCase, deadline)
+          }
+      }
+  }
+}
+
+/**
+ * Provide a class to control a Spark application using YARN API.
+ *
+ * @param appIdFuture A future that returns the YARN application id for this application.
+ * @param process The spark-submit process launched the YARN application. This is optional.
+ *                If it's provided, SparkYarnApp.log() will include its log.
+ */
+class SparkYarnApp (
+    appIdFuture: Future[ApplicationId],
+    process: Option[LineBufferedProcess],
+    listener: Option[SparkAppListener],
+    yarnClient: => YarnClient = SparkYarnApp.getYarnClient()) // For unit test.
+  extends SparkApp
+  with Logging {
+  import SparkYarnApp.ec
+
+  private var state: SparkApp.State = SparkApp.State.STARTING
+  private var yarnDiagnostics: IndexedSeq[String] = IndexedSeq.empty[String]
+
+  override def log(): IndexedSeq[String] =
+    (process map (_.inputLines) getOrElse ArrayBuffer.empty[String]) ++ yarnDiagnostics
+
+  override def kill(): Unit = synchronized {
+    if (isRunning) {
+      process foreach (_.destroy())
```

Comment:
Shouldn't you do this after you try to kill the application through YARN? Otherwise you might be making the timeout case more common by killing the spark-submit process mid-submission.